content/blog/2020-12-20-loop-flatten/index.md

```diff
@@ -0,0 +1,199 @@
+  for (int j = 0; j < m; j+= 2) { // unrolled inner loop
+    Inner_Stmt(i);
+    Inner_Stmt(i+1);
+  }
+}
+```
+
+3. Loop flattening<br/>
+As illustrated by its name, this optimization will flatten the nested loop into a single-level loop. 
+To be more specific, it will transform the following piece of code:
+```javascript
+  for (int i = 0; i < n; i++) { // unflattened nested loop
+    for (int j = 0; j < m; j++) {
+      Stmt(f(i, j))
+  }
+}
+```
+Into this piece of code:
+
+```javascript
+for (int i = 0; i < n*m; i+= 2) { // flattened loop
+  Stmt(f(i))
+}
+```
+
+After flattening the loop, the loop bound becomes larger so that we can apply larger unrolling parameters to further explore the parallelism. Also, this technique is necessary for some optimizations in High-Level Synthesis tools. For example, suppose we have a nested for loop with variable loop bound, the tool cannot automatically apply the pipeline optimization because it has no knowledge about how to deal with the inner loop. Thus we need to manually flatten the nested loop and then apply the pipeline pragma. To learn more about the importance of the loop flatten technique for pipelining, please check out [this link](https://people.ece.uw.edu/hauck/publications/LoopFlattening.pdf).
+
+
+## Implementation
+The idea is quite straightforward but the implementation is not that easy. To implement this pass, we should perform the following modifications.
+
+1. Update the loop bound and increment <br/>
+As the first step, we should update the loop bound for the outer loop. Suppose the loop bound for the inner loop and outer loop are m and n with the same increment, we can change the loop bound for the outer loop to be `m*n` and keep the same increment.
+
+2. Extract the statements inside the inner loop to the outer loop<br/>
+In this step, we extract all the statements inside the inner loop and put them into the outer loop. At the same time, we should detect the usage of the inner loop variable and then remove it with the correct one after flattening the loop. For example, suppose we have a piece of code like this inside the inner loop: `i*n+j`, whereas `i` and `j` are the loop variables, `n` is the outer loop bound, and the increment for both inner loop and outer loop is 1. To correctly flatten this loop, we should replace the statement `i*n+j` with `i`. 
+
+3. Delete the inner loop<br/>
+After replacing the statement from the inner loop to the outer loop, we can safely remove the inner loop by deleting the corresponding basic blocks.
+
+Finally, we should check the equivalence of the code, i.e., the flattened loop should perform the same execution as the original nested loop. This can be verified by testing the pass with benchmarks. 
+
+During the implementation, I got stuck in the second step, i.e., extract the statements inside the inner loop to the outer loop. The generated code performs weird when extracts the loop variant expressions from the inner loop to the outer loop. 

```

Comment:
What exactly goes wrong? Some specificity here about what "weird" thing happened would be useful.